### PR TITLE
Factoring Trinomials and Difference of Squares

### PR DESCRIPTION
**Pull request by aliang8**
`PERFECT_SQUARE_TRINOMIAL` and `FACTOR_TRINOMIAL` logic is somewhat redundant, not sure if we should remove it and just have one rule. Code should be refactored if possible.

### CODE REVIEW

File: lib/rules.js [hunk 1]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+// assume monic polynomial
+// TODO: add min and max to evaluate
+
+export const FACTOR_SYMBOL = defineRuleString('#a^#b_0 + ...', '#a^#eval(min(#b_0, ...)) (#a^#eval(#b_0 - min(#b_0, ...)) + ...)', {b: query.isNumber})
+
+// TODO: add restrictions (#c_0 divisible by 2 && #a_0 is perfect square)
+
+// e.g. 4x^2 - 9y^2 -> (2x)^2 - (3y)^2
+export const FACTOR_DIFFERENCE_OF_SQUARES_HELPER =
+    defineRuleString('#a_0 #b_0^#c_0 - #d_0 #e_0^#f_0', '(#eval(nthRoot(#a_0)) #b_0^(#eval(#c_0/2)))^2 - (#eval(nthRoot(#d_0)) #e_0^(#eval(#f_0/2)))^2')
```
Comment (k4b7) [inline lib/rules.js]: The subscripts aren't necessary here because there are `...`

File: lib/rules.js [hunk 2]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+
+// e.g. (2x)^2 - (3y)^2 -> (2x + 3y)(2x - 3y)
+export const FACTOR_DIFFERENCE_OF_SQUARES =
+    defineRuleString('#a^2 - #b^2', '(#a + #b)(#a - #b)')
+
+
+export const nthRoot = (num, root = 2, precision = 12) => {
```
Comment (k4b7) [inline lib/rules.js]: We might want to pull this out into `math-evaluator` so that we can call it from within `#eval(...)`.

File: lib/rules.js [hunk 3]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+
+// TODO: add restrictions (#c_0 divisible by 2 && #a_0 is perfect square)
+
+// e.g. 4x^2 - 9y^2 -> (2x)^2 - (3y)^2
+export const FACTOR_DIFFERENCE_OF_SQUARES_HELPER =
+    defineRuleString('#a_0 #b_0^#c_0 - #d_0 #e_0^#f_0', '(#eval(nthRoot(#a_0)) #b_0^(#eval(#c_0/2)))^2 - (#eval(nthRoot(#d_0)) #e_0^(#eval(#f_0/2)))^2')
+
+// e.g. (2x)^2 - (3y)^2 -> (2x + 3y)(2x - 3y)
+export const FACTOR_DIFFERENCE_OF_SQUARES =
+    defineRuleString('#a^2 - #b^2', '(#a + #b)(#a - #b)')
```
Comment (k4b7) [inline lib/rules.js]: nice

File: lib/rules.js [hunk 4]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+    (node) => {
+        let isTrinomial = false
+        let isPerfectSquareTrinomial = false
+        // trinomial is a perfect square if it satisfies the conditions that
+        // a^2 + 2ab + b^2 = (a + b)^2
+        // 1. First & third term are perfect squares
+        // 2. Middle coefficient is 2 / -2 * (sqrt(a^2)) * (sqrt(b^2))
+        if (query.isAdd(node) && node.args.length === 3) {
```
Comment (k4b7) [inline lib/rules.js]: This means we wont be able to factor trinomials that are inside a larger expression, e.g.
```
1 + 4x^2 + 12x + 9 + x
```

File: lib/rules.js [hunk 5]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+            const [firstTerm, secondTerm, thirdTerm] = node.args
+
+            isTrinomial =
+                isPolynomialTerm(firstTerm)
+                && isPolynomialTerm(secondTerm)
+                && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm) || canApplyRule(CANCEL_EXPONENT, parse(`nthRoot(${print(thirdTerm)})`)))
+            if (isTrinomial) {
+                const firstCoef = query.getValue(firstTerm.args[0])
+                const firstPoly = firstTerm.args[1]
+                const firstDegree = query.getValue(firstPoly.args[1])
```
Comment (k4b7) [inline lib/rules.js]: This approach will leave out lots of of stuff, e.g. we won't be able to factor things like:
```
4x^2y^2 + 12a^2b^2xy + 9x^b^4
```
We'll probably want to leverage some of the helper functions from https://github.com/semantic-math/math-rules/blob/master/lib/rules/collect-like-terms.js to make this more robust.

File: lib/rules.js [hunk 6]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+    },
+
+    (node) => {
+        const [firstTerm, secondTerm, thirdTerm] = node.args
+        const firstCoef = query.getValue(firstTerm.args[0])
+        const firstPoly = firstTerm.args[1]
+        const firstDegree = query.getValue(firstPoly.args[1])       
+        const thirdCoef = isPolynomialTerm(thirdTerm) && !query.isNumber(thirdTerm)
+              ? query.getValue(thirdTerm.args[0])
+              : query.getValue(thirdTerm)
```
Comment (k4b7) [inline lib/rules.js]: Use helpers from collect-like-terms.js, e.g. `getCoefficient`

File: lib/rules.js [hunk 7]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+
+        const firstFactor = parse(`${Math.sqrt(firstCoef)} ${getVariableFactorName(firstPoly)}^${firstDegree/2}`)
+        const secondFactor = thirdPoly
+              ? `${Math.sqrt(thirdCoef)} ${getVariableFactorName(thirdPoly)}^${thirdDegree/2}`
+              : Math.sqrt(query.getValue(thirdTerm))
+
+        const result = query.isNeg(secondTerm)
+              ? parse(`(${print(firstFactor)} - ${secondFactor})^2`)
+              : parse(`(${print(firstFactor)} + ${secondFactor})^2`)
+        return result
```
Comment (k4b7) [inline lib/rules.js]: It would be nice to do this using the builders... but they're kind of unwieldily at the moment.  Can you add a TODO?

Comment (k4b7) [reply lib/rules.js]: I've update `math-nodes` to have more build functions, see https://github.com/semantic-math/math-nodes/commit/9dbd3e71bef309b0e406b19a5c152d5f1952fd2a.

This would become:
```
const result = query.isNeg(secondTerm)
   ? build.add(clone(firstFactor), clone(secondFactor))
   : build.sub(clone(firstFactor), clone(secondFactor))
```

File: lib/rules.js [hunk 8]
```diff
@@ -353,6 +353,338 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+    }
+)
+
+function getFactors(num){
```
Comment (k4b7) [inline lib/rules.js]: Can you add a comment stating how this differs from `primeFactorization`?

**Review by k4b7 — CHANGES_REQUESTED** [review]

File: lib/rules.js (hunk 1 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 8 above)
Comment (aliang8) [reply lib/rules.js]: Hm, I guess there is no difference. It just formatted in a friendly way to be used in the rule.

Comment (k4b7) [reply lib/rules.js]: There's a slightly difference though in the output... you remove duplicate factors right?

File: lib/rules.js [hunk 9]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
 // e.g nthRoot(9, 2) -> 3
 export const NTH_ROOT_VALUE = defineRuleString('nthRoot(#a, #b)', '#eval(nthRoot(#a, #b))', {a: query.isNumber, b: query.isNumber})
 
+// FACTOR
+
+// TODO: add minus case
+// assume monic polynomial
+// TODO: add min and max to evaluate
+
+export const FACTOR_SYMBOL = defineRuleString('#a^#b_0 + ...', '#a^#eval(min(#b_0, ...)) (#a^#eval(#b_0 - min(#b_0, ...)) + ...)', {b: query.isNumber})
```
Comment (k4b7) [inline lib/rules.js]: Neato!  Could you provide an example what this does?

File: lib/rules.js [hunk 10]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+
+// TODO: handle fractional coefficients
+// TODO: isPolynomialTerm should handle fractional coeffs
+
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+    (node) => {
+        const isFactorable = canApplyRule(FACTOR_SUM_PRODUCT_RULE, node)
+        if (isFactorable) {
+            const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
```
Comment (k4b7) [inline lib/rules.js]: Nice!

File: lib/rules.js [hunk 11]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
 // e.g nthRoot(9, 2) -> 3
 export const NTH_ROOT_VALUE = defineRuleString('nthRoot(#a, #b)', '#eval(nthRoot(#a, #b))', {a: query.isNumber, b: query.isNumber})
 
+// FACTOR
```
Comment (k4b7) [inline lib/rules.js]: Can you put these in their own file?  I'd like eventually rename this file to be `simple-rules.js` and have it contain only rules which can be full expressed with patterns.

File: lib/rules.js [hunk 12]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+            const isTrinomial =
+                  isPolynomialTerm(firstTerm)
+                  && isPolynomialTerm(secondTerm)
+                  && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm))
+
+            if (isTrinomial) {
+                let isFactorable = false
+                const {constants, coefficientMap} = getCoefficientsAndConstants(node)
+                const variables = Object.keys(coefficientMap).map(key => parse(key))
+                const coefficients = Object.keys(coefficientMap).map(key => {return coefficientMap[key]})
```
Comment (k4b7) [inline lib/rules.js]:
```
const coeffs = Object.keys(coeffMap).map(key => coeffMap[key][0])
```
This keeps it to a single line and you don't have to do `coeffs[1][0]` later you can just do `coeffs[1]`.

File: lib/rules.js [hunk 13]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+    (node) => {
+        if (query.isAdd(node) && node.args.length === 3) {
+            const [firstTerm, secondTerm, thirdTerm] = node.args
+
+            // First two terms should be polynomials
+            // Third term is either number or polynomial
+            const isTrinomial =
+                  isPolynomialTerm(firstTerm)
+                  && isPolynomialTerm(secondTerm)
+                  && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm))
```
Comment (k4b7) [inline lib/rules.js]: We should probably handle things like `1 + 2x + x^2` as well... maybe we should have a transform to order terms in a polynomial in descending order.  This can be a TODO.

File: lib/rules.js [hunk 14]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                // Second polynomial to be matched
+                const firstMiddle = firstPoly.map(term => {
+                    const identifier = term.args[0]
+                    const exponent = build.number(query.getValue(term.args[1]) / 2)
+                    return build.applyNode('pow', [identifier, exponent])
+                })
+
+                let secondMiddle
+                if (thirdPoly) {
+                    secondMiddle = thirdPoly.map(term => {
```
Comment (k4b7) [inline lib/rules.js]: rename `term` to `factor`

File: lib/rules.js [hunk 15]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                const thirdPoly =
+                      variables[2]
+                      ? query.isMul(variables[2])
+                      ? variables[2].args
+                      : [variables[2]]
+                      : null
+
+                const secondCoef = coefficients[1][0]
+                // Second polynomial to be matched
+                const firstMiddle = firstPoly.map(term => {
```
Comment (k4b7) [inline lib/rules.js]: rename `term` to `factor`

File: lib/rules.js [hunk 16]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                    const identifier = term.args[0]
+                    const exponent = build.number(query.getValue(term.args[1]) / 2)
+                    return build.applyNode('pow', [identifier, exponent])
+                })
+
+                let secondMiddle
```
Comment (k4b7) [inline lib/rules.js]: Can you provide an explanation of what `firstMiddle` and `secondMiddle` are and what the algorithm is at a high level?

File: lib/rules.js [hunk 17]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                        const identifier = term.args[0]
+                        const exponent = build.number(query.getValue(term.args[1]) / 2)
+                        return build.applyNode('pow', [identifier, exponent])
+                    })
+                }
+
+                const matchSecondTerm =  secondMiddle
+                      ? build.applyNode ('mul',
+                                         [secondCoef, ...firstMiddle, ...secondMiddle]
+                                         , {implicit: true})
```
Comment (k4b7) [inline lib/rules.js]: This can be
```
build.implicitMul(secondCoeff, ...firstMiddle, ...secondMiddle)
```

File: lib/rules.js [hunk 18]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                    const exponent = build.number(query.getValue(term.args[1]) / 2)
+                    return build.applyNode('pow', [identifier, exponent])
+                })
+
+                let secondMiddle
+                if (thirdPoly) {
+                    secondMiddle = thirdPoly.map(term => {
+                        const identifier = term.args[0]
+                        const exponent = build.number(query.getValue(term.args[1]) / 2)
+                        return build.applyNode('pow', [identifier, exponent])
```
Comment (k4b7) [inline lib/rules.js]: switch to `build.pow(identifier, exponent)`

File: lib/rules.js [hunk 19]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                      : build.applyNode ('mul',
+                                         [secondCoef, ...firstMiddle]
+                                         , {implicit: true})
+
+                // General conditions:
+                // All degrees in 1st and 3rd term must be even
+                // secondTerm = matchSecondTerm
+                isFactorable =
+                    firstPoly.forEach(term => {return term.args[1] % 2 == 0})
+                    && thirdPoly ? thirdPoly.forEach(term => {return term.args[1] % 2 == 0}) : true
```
Comment (k4b7) [inline lib/rules.js]: The `return` isn't necessary here.  Also, I think you meant `every` not `forEach`.

File: lib/rules.js [hunk 20]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                    && print(secondTerm) == print(flattenOperands(matchSecondTerm))
+                return isFactorable ? {node} : null
+            }
+        }
+    },
+
+    (node) => {
+        const {constants, coefficientMap} = getCoefficientsAndConstants(node)
+        const variables = Object.keys(coefficientMap).map(key => parse(key))
+        const coefficients = Object.keys(coefficientMap).map(key => {return coefficientMap[key]})
```
Comment (k4b7) [inline lib/rules.js]: `return` isn't necessary.

File: lib/rules.js [hunk 21]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+
+        const firstFactors = getFactors(firstCoef)
+        const thirdFactors = getFactors(Math.abs(thirdCoef))
+
+        let combo
+
+        for (var i in firstFactors){
+            for (var j in thirdFactors) {
+                const l1 = firstFactors[i]
+                const l2 = thirdFactors[j]
```
Comment (k4b7) [inline lib/rules.js]: I have no idea what's going on here.  Can you add a comment explaining this algorithm?

File: lib/rules.js [hunk 22]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+        }
+
+        const result = build.applyNode
+        ('mul',
+         combo.map(factor => {
+             const firstFactor =
+                   build.apply(
+                       'mul',
+                       [build.number(factor[0]), ...firstPoly.map(
+                           term =>
```
Comment (k4b7) [inline lib/rules.js]: `term -> factor`

File: lib/rules.js [hunk 23]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                    combo = []
+                }
+            }
+        }
+        if (combo == []) {
+            throw new Error('cannot factor polynomial')
+        }
+
+        const result = build.applyNode
+        ('mul',
```
Comment (k4b7) [inline lib/rules.js]: All of the these implicit `mul`s can be build using `build.implicitMul`.

File: lib/rules.js [hunk 24]
```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
...
+                } else if (eight == secondCoef && l2[0] * -l2[1] == thirdCoef){
+                    combo = [[l1[0],-l2[1]],[l1[1],l2[0]]]
+                } else {
+                    combo = []
+                }
+            }
+        }
+        if (combo == []) {
+            throw new Error('cannot factor polynomial')
+        }
```
Comment (k4b7) [inline lib/rules.js]: If feels weird that `canApplyRule` might be successful, but `applyRule` might still fail.

File: lib/rules.js (hunk 17 above)
Comment (k4b7) [reply lib/rules.js]: You may need to update `math-nodes`.

**Review by k4b7 — CHANGES_REQUESTED** [review]

File: lib/rules.js (hunk 9 above)
Comment (aliang8) [reply lib/rules.js]: Ahhh, this doesn't exactly work yet. This rule only applies to addition (x^2 + x^3 + x^4 -> x^2(x^0 + x^1 + x^2). Not exactly sure if we want two separate rules for addition and subtraction or if there is a way to do both in a rule string. I also need the min and max functions to be exposed in math-evaluator.

File: lib/rules.js (hunk 13 above)
Comment (aliang8) [reply lib/rules.js]: Yup I was thinking to go with the second approach.

File: lib/rules.js (hunk 24 above)
Comment (aliang8) [reply lib/rules.js]: Good point, I have to fix this.

Comment (aliang8) [reply lib/rules.js]: Not entirely sure how to without replicating the code for `rewriteFn`.

File: lib/rules.js (hunk 9 above)
Comment (k4b7) [reply lib/rules.js]: I think this should work with subtraction as is.  Have you tried writing a test case with subtraction?

File: lib/rules.js (hunk 24 above)
Comment (k4b7) [reply lib/rules.js]: I would extract a common function.  Obviously it's not ideal to run the same code twice, but I'd rather maintain the invariant that if `canApplyRule` returns `true` that `applyRule` will succeed.

File: lib/rules.js (hunk 9 above)
Comment (aliang8) [reply lib/rules.js]: Yeah, I need min and max to support nary parameters.

File: lib/factor-rules.js [hunk 25]
```diff
@@ -0,0 +1,373 @@
...
+
+  (node) => {
+    const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    return build.pow(result.args[0], build.number(2))
+  }
+)
+
+const getExponent = (node) => {
+    if (query.isNumber(node)) {
+        return 0
```
Comment (k4b7) [inline lib/factor-rules.js]: This seems like a weird value to return for a number.  I would expect the exponent to be `1`.  Maybe there should be an `if` statement before calling `getExponent` to check if the node is a number.

File: lib/factor-rules.js [hunk 26]
```diff
@@ -0,0 +1,373 @@
...
+    } else if (query.isIdentifier(node)){
+        return 1
+    } else if (query.isPow(node)) {
+        return query.getValue(node.args[1])
+    } else if (query.isMul(node)){
+        return getExponent(node.args[1])
```
Comment (k4b7) [inline lib/factor-rules.js:121]: Shouldn't this be `return Math.max(...node.args.map(getExponent))`?

File: lib/factor-rules.js [hunk 27]
```diff
@@ -0,0 +1,373 @@
...
+    } else if (query.isNeg(node)) {
+        const variable = node.args[0]
+        return getExponent(variable.args[1])
+    } else {
+        return null
+    }
+}
+
+// e.g. 2 + 3x^2 + 3x - 4x^3 -> -4x^3 + 3x^2 + 3x + 2
+export const REARRANGE_TERMS = defineRule(
```
Comment (k4b7) [inline lib/factor-rules.js]: This doesn't take into account polynomials with multiple variables.  Maybe add a TODO.

File: lib/factor-rules.js (hunk 26 above)
Comment (k4b7) [reply lib/factor-rules.js:121]: It's okay to limit ourselves to single variables for now, but we should have a TODO somewhere.

File: lib/factor-rules.js [hunk 28]
```diff
@@ -0,0 +1,373 @@
...
+  1. Get factors for 6 and 2
+  6: [1,6], [2,3]    2: [1,2]
+
+  2. For each pair of factors (one from the first
+  and one from the last term) ...
+  [1,6] and [1,2]      |       [2,3] and [1,2]
+
+  3. Cross multiply the factors and check if
+  the result matches one of the eight conditions
+  (the second factor shld also multiply to equal the thirdCoef)
```
Comment (k4b7) [inline lib/factor-rules.js]: shld --> should

File: lib/factor-rules.js [hunk 29]
```diff
@@ -0,0 +1,373 @@
...
+  [1,6] and [1,2]
+
+  Check: 1 * 2 + 6 * 1 ?= 7 so on and so forth
+  Once we reach 2 * 2 + 3 * 1 ?= 7, we have found a match.
+  (Note: 3 * 1 = 3 satisfying the second condition)
```
Comment (k4b7) [inline lib/factor-rules.js:178]: I'm confused about this note.... don't we want to check if the multiple the first candidate coefficients results in the first coefficient of the trinomial and multiplying the both of the second candidate numbers results in the constant of the trinomial.  In this case 2 * 3 = 6 and 2 * 1 = 2 so it works out.

File: lib/factor-rules.js [hunk 30]
```diff
@@ -0,0 +1,373 @@
...
+        }
+    },
+
+    (node) => {
+        const ordered = node.args.sort(function(a,b) {return getExponent(b) - getExponent(a)})
+        return build.add(...ordered)
+    }
+)
+
+const getFactors = (num) => {
```
Comment (k4b7) [inline lib/factor-rules.js]: This should maybe be called `getFactorPairs` or add a comment describing what it does.  I was confused b/c I thought calling `getFactors(12)` would return `[1, 2, 3, 4, 6, 12]` but in fact it returns `[[1, 12], [2, 6], [3, 4]]`.

File: lib/factor-rules.js [hunk 31]
```diff
@@ -0,0 +1,373 @@
...
+  const firstPoly = query.isMul(variables[0]) ? variables[0].args : [variables[0]]
+
+  const secondCoef = query.getValue(coeffs[1])
+
+  const thirdPoly =
+        variables[2]
+        ? query.isMul(variables[2])
+        ? variables[2].args
+        : [variables[2]]
+        : null
```
Comment (k4b7) [inline lib/factor-rules.js]: Can you rewrite this ternary as an if-else?

File: lib/factor-rules.js [hunk 32]
```diff
@@ -0,0 +1,373 @@
...
+
+  const thirdCoef =
+        thirdPoly ?
+        query.getValue(coeffs[2])
+    : constants[0]
+    ? query.getValue(constants[0])
+    : null
```
Comment (k4b7) [inline lib/factor-rules.js]: same

File: lib/factor-rules.js [hunk 33]
```diff
@@ -0,0 +1,373 @@
...
+  [1,6] and [1,2]
+
+  Check: 1 * 2 + 6 * 1 ?= 7 so on and so forth
+  Once we reach 2 * 2 + 3 * 1 ?= 7, we have found a match.
+  (Note: 3 * 1 = 3 satisfying the second condition)
+
+  At this point, we find that the correct combination is
+  : [2,3] and [1,2] -> (2x + 3)(1x + 2)
+*/
+const factor_trinomial_helper = (node) => {
```
Comment (k4b7) [inline lib/factor-rules.js]: Does this handle cases where the lead coefficient is negative?

File: lib/factor-rules.js [hunk 34]
```diff
@@ -0,0 +1,373 @@
...
+        const variables = Object.keys(coefficientMap).map(key => parse(key))
+        const coeffs = Object.keys(coefficientMap).map(key => coefficientMap[key][0])
+
+        const firstPoly = query.isMul(variables[0]) ? variables[0].args : [variables[0]]
+        const thirdPoly =
+              variables[2]
+              ? query.isMul(variables[2])
+              ? variables[2].args
+          : [variables[2]]
+        : null
```
Comment (k4b7) [inline lib/factor-rules.js]: The indentation is a little off here.  It should probably be:
```
const thirdPoly =
    variables[2]
       ? query.isMul(variables[2])
            ? variables[2].args 
            : [variables[2]]
       : null
```
even that is hard to parse... since the default is `null` maybe do:
```
let thirdPoly = null;
if (variables[2]) {
   if (query.isMul(variables[2]) {
      thirdPoly = variables[2].args
   } else {
      thirdPoly = [variables[2]]
   }
}
```

File: lib/factor-rules.js [hunk 35]
```diff
@@ -0,0 +1,373 @@
...
+    const variables = Object.keys(coefficientMap).map(key => parse(key))
+
+    const firstPoly = query.isMul(variables[0]) ? variables[0].args : [variables[0]]
+
+    const thirdPoly =
+          variables[2]
+      ? query.isMul(variables[2])
+          ? variables[2].args
+      : [variables[2]]
+          : null
```
Comment (k4b7) [inline lib/factor-rules.js]: Please rewrite as `if-else` statements.

File: lib/factor-rules.js (hunk 25 above)
Comment (aliang8) [reply lib/factor-rules.js]: Oh, the `getExponent` function is to get the degree of x for a given polynomial term. So for instance, `6x^2 + 3x + 1`, `getExponent` on 1 would return 0 because it is `1x^0`. Maybe the function name itself is a bit misleading?

File: lib/factor-rules.js (hunk 29 above)
Comment (aliang8) [reply lib/factor-rules.js:178]: You already know that `2 * 3 = 6` and `2 * 1 = 2`. It comes from the fact that those are the factors of 6 and 2 respectively (pairs that we got from `firstFactors` and `thirdFactors`). The note here is just to make sure we don't have an erroneous combo / solution (this happened a few times in my test cases).

File: lib/factor-rules.js (hunk 30 above)
Comment (aliang8) [reply lib/factor-rules.js]: Sorry!

File: lib/factor-rules.js (hunk 33 above)
Comment (aliang8) [reply lib/factor-rules.js]: We assume that lead coefficient is positive because it should have been factored out previously by another rule.

File: lib/factor-rules.js (hunk 29 above)
Comment (k4b7) [reply lib/factor-rules.js:178]: I need to spend some more time with this to really understand what's going on.  I have some of pieces but not everything.  This is eventually going to get re-reviewed when it gets moved over to mathsteps so I'm okay with merging this.

**Review by k4b7 — APPROVED** [review]

File: lib/factor-rules.js (hunk 29 above)
Comment (aliang8) [reply lib/factor-rules.js:178]: Awesome, let me know when you merge this, so I can update my other PRs.
